cli/factory.rs

```diff
@@ -367,6 +367,19 @@ impl CliFactory {
             }),
           })
         } else {
+          let maybe_node_modules_path = cli_options
+            .node_modules_dir_path()
+            .and_then(|node_modules| if let Ok(deno_dir) = self.deno_dir() {
+              if node_modules.starts_with(&deno_dir.root) {
+                // If the node_modules path is inside the deno_dir, we don't want to use it.
+                None
+              } else {
+                Some(node_modules.clone())
+              }
+            } else {
+              Some(node_modules.clone())
+            });
```

Comment:
As discussed in person: instead of doing this here, let's move this up into the `resolve_node_modules_folder` function in cli/args/mod.rs (https://github.com/denoland/deno/blob/d1b2a9822d499def513c5b178927c4e05c39caff/cli/args/mod.rs#L1746)

This requires having access to the deno dir at this point though. I think we could:

1. Move the `DenoDirProvider` struct out of cli/factory.rs and instead create it within `CliOptions::new` and store it on `CliOptions`.
2. Expose `DenoDirProvider` on `CliOptions`.
3. Pass `DenoDirProvider` into the `resolve_node_modules_folder` function and then use it in that function.
4. Get and canonicalize the deno dir path inside `resolve_node_modules_folder` in order to compare it against the canonicalized node_modules dir path.